File: mercury/galois_solve.m
%---------------------------------------------------------------------------%
% vim: ft=mercury ts=4 sw=4 et
%---------------------------------------------------------------------------%
% File: galois_solve.m
% Main author: nyc
%
% Top-level interface for solving polynomials via Galois group computation.
%
% Composes:
%   - galois_identify: identifies the Galois group of an irreducible
%     polynomial (currently degree 5 only)
%   - radical_tower: constructs an explicit radical tower and solves
%     via Lagrange resolvents
%   - rad_alg_convert: handles degrees 1–4 via quadratic/Cardano/Ferrari
%
% For degree ≤ 4, solve_alg_num returns no so that the caller can use
% direct formulas. For degree 5, the Galois pipeline is invoked.
%
%---------------------------------------------------------------------------%

:- module galois_solve.
:- interface.

:- import_module alg_num.
:- import_module list.
:- import_module maybe.
:- import_module poly.
:- import_module rad_expr.
:- import_module rational.

%---------------------------------------------------------------------------%

    % Try to express an algebraic number as a radical expression.
    %
    % Returns no for degree ≤ 4 (caller should use Cardano/Ferrari),
    % for degree > 5 (not yet supported), or for non-solvable groups.
    %
:- func solve_alg_num(alg_num) = maybe(rad_expr(rational)).

    % Solve an irreducible polynomial, returning radical expressions
    % for all roots.
    %
:- func solve_poly(poly(rational)) = maybe(list(rad_expr(rational))).

    % Identify the Galois group and solve if solvable.
    % Returns {group_name, roots} or no.
    %
:- func identify_and_solve(poly(rational)) =
    maybe({string, list(rad_expr(rational))}).

    % Pick the radical expression whose numerical value is closest to
    % the target real value.
    %
:- func pick_closest_real(list(rad_expr(rational)), float) =
    maybe(rad_expr(rational)).

%---------------------------------------------------------------------------%
%---------------------------------------------------------------------------%

:- implementation.

:- import_module bool.
:- use_module float.
:- use_module int.
:- use_module math.
:- import_module galois_identify.
:- import_module integer.
:- import_module rad_eval.
:- import_module radical_tower.
:- import_module string.
:- import_module transitive_group.

%---------------------------------------------------------------------------%

solve_alg_num(A) = Result :-
    P = alg_min_poly(A),
    D = poly.degree(P),
    ( if int.'=<'(D, 4) then
        Result = no
    else
        ( if solve_poly(P) = yes(AllRoots) then
            ApproxVal = rational_to_float(
                alg_approx(from_integers(integer.one,
                                         integer(1000000000000000)), A)),
            Result = pick_closest_real(AllRoots, ApproxVal)
        else
            Result = no
        )
    ).

solve_poly(F) = Result :-
    ( if identify_and_solve(F) = yes({_, Roots}) then
        Result = yes(Roots)
    else
        Result = no
    ).

identify_and_solve(F) = Result :-
    ( if poly.degree(F) \= 5 then
        Result = no
    else
        ( if identify_galois_group_5(F) = yes(GR) then
            TG = gr_group(GR),
            ( if tg_solvable(TG) = bool.yes then
                ( if solve_via_tower(GR, F) = yes(Roots) then
                    Result = yes({tg_name(TG), Roots})
                else
                    Result = no
                )
            else
                Result = no
            )
        else
            Result = no
        )
    ).

%---------------------------------------------------------------------------%
% Pick closest real root
%---------------------------------------------------------------------------%

pick_closest_real(Exprs, Target) = Result :-
    Scored = list.map(
        func(E) = {E, eval_dist(E, Target)},
        Exprs),
    Valid = list.filter(
        (pred({_, D}::in) is semidet :- float.'<'(D, 1.0e-4)),
        Scored),
    ( if Valid = [_ | _] then
        Result = yes(fst(find_min_pair(Valid)))
    else
        Best = find_min_pair(Scored),
        ( if float.'<'(snd(Best), 1.0) then
            Result = yes(fst(Best))
        else
            Result = no
        )
    ).

:- func eval_dist(rad_expr(rational), float) = float.

eval_dist(E, Target) = Result :-
    V = eval_complex(E),
    ( if float.'<'(float.abs(im(V)), 1.0e-6) then
        Result = float.abs(float.'-'(re(V), Target))
    else
        Dre = float.'-'(re(V), Target),
        Dim = im(V),
        Result = math.sqrt(float.'+'(float.'*'(Dre, Dre),
                                      float.'*'(Dim, Dim)))
    ).

:- func find_min_pair(list({rad_expr(rational), float})) =
    {rad_expr(rational), float}.

find_min_pair(Xs) = Result :-
    ( if Xs = [H | T] then
        Result = list.foldl(
            func({E, D}, {BE, BD}) = R :-
                ( if float.'<'(D, BD) then R = {E, D}
                else R = {BE, BD} ),
            T, H)
    else
        % Should not happen — caller ensures non-empty
        Result = {re_lit(rational.zero), 1.0e10}
    ).

:- func fst({A, B}) = A.
fst({X, _}) = X.

:- func snd({A, B}) = B.
snd({_, Y}) = Y.

:- func rational_to_float(rational) = float.

rational_to_float(R) = Result :-
    N = integer.to_string(numer(R)),
    D = integer.to_string(denom(R)),
    ( if string.to_float(N, NF), string.to_float(D, DF) then
        Result = float.'/'(NF, DF)
    else
        Result = 0.0
    ).

%---------------------------------------------------------------------------%
:- end_module galois_solve.
%---------------------------------------------------------------------------%

File: ocaml/lib/galois_solve.ml
(** Radical solutions for solvable polynomials.

    Top-level interface for solving polynomials via Galois group
    computation. Composes Identify (group identification) and
    Radical_tower (Lagrange resolvent descent).

    For degree <= 4, returns [None] so the caller can use direct
    Cardano/Ferrari formulas. For prime degree >= 5, identifies
    the Galois group and solves if solvable. *)

module P = Poly.RatPoly

(** Identify the Galois group and solve if solvable.

    Returns [Some (group_name, roots)] if the polynomial has a
    solvable Galois group and can be expressed in radicals.
    Returns [None] if the group is non-solvable (A_n or S_n),
    the degree is unsupported, or solving fails. *)
let identify_and_solve f =
  if P.degree f <= 4 then None
  else
    match Identify.identify f with
    | None -> None
    | Some gr ->
      if not gr.Identify.group.Transitive_group.solvable then None
      else
        match Radical_tower.solve_via_tower gr f with
        | None -> None
        | Some roots ->
          Some (Identify.group_name gr, roots)

(** Attempt to solve a polynomial by radicals.
    Returns [Some roots] if the Galois group is solvable,
    [None] otherwise. *)
let solve f : Rational.t Rad_expr.t list option =
  match identify_and_solve f with
  | None -> None
  | Some (_, roots) -> Some roots

(** Check if a polynomial is solvable by radicals. *)
let is_solvable f =
  if P.degree f <= 4 then true
  else
    match Identify.identify f with
    | None -> false
    | Some gr -> gr.Identify.group.Transitive_group.solvable

(** Pick the radical expression whose numerical value is closest
    to a target real value. *)
let pick_closest_real exprs target =
  let scored = List.map (fun e ->
    let v = Eval.eval_complex e in
    let dist =
      if Float.abs v.Complex.im < 1e-6 then
        Float.abs (v.Complex.re -. target)
      else
        Complex.norm (Complex.sub v { Complex.re = target; im = 0.0 })
    in
    (e, dist)
  ) exprs in
  let valid = List.filter (fun (_, d) -> d < 1e-4) scored in
  match valid with
  | [] ->
    let (best_e, best_d) = List.fold_left (fun (be, bd) (e, d) ->
      if d < bd then (e, d) else (be, bd)
    ) (List.hd scored) scored in
    if best_d < 1.0 then Some best_e else None
  | _ ->
    Some (fst (List.fold_left (fun (be, bd) (e, d) ->
      if d < bd then (e, d) else (be, bd)
    ) (List.hd valid) valid))
